### PR TITLE
fix(CLOUDDST-30777): log improvements for deploy-fbc-operator

### PR DESCRIPTION
**Pull request by JAVGan**
This PR contains the following changes:

## Validate parameters for deploy-fbc-operator
    
This commit introduces a new task named `validate-parameters` which will  be executed first in the pipeline.
    
This newly added task will raise errors when required secrets/parameters are not set as well as emit warnings for the ones which are missing so the end user can properly debug.
    
The goal of this change is to prevent a scenario which the pipeline gets stuck without any error messages due to missing secrets, as well as early failing/warning for missing fields.

## Log skipped tasks for deploy-fbc-operator
    
This commit enhances the logging for `deploy-fbc-operator` by informing which tasks are going to be skipped for non "push" events, such as  "pull_request" or "retest-all-comment".

## JIRA
Refers to CLOUDDST-30777

### CODE REVIEW

**Comment by JAVGan** [comment]
@nmars can you please check whether this makes sense?

**Comment by JAVGan** [comment]
@nmars @yashvardhannanavati I also updated the `1.0` for the sake of "why not" and it's raising errors on `Validate Tasks and Pipelines YAMLs` but it's unrelated to my change AFIAK

File: pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml [hunk 1]
```diff
@@ -89,7 +89,85 @@ spec:
       default: "https://konflux-ui.apps.stone-prd-rh01.pg1f.p1.openshiftapps.com"
       type: string
   tasks:
+    - name: validate-parameters
+      taskSpec:
+        steps:
+          - name: validate-parameters
+            image: quay.io/konflux-ci/konflux-test:v1.4.48@sha256:9b815268fb2bf10b5d745518da1c6568944f15816efe51adc192972b42a6e74d
```
Comment (nmars) [inline pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml]: I suggest we use the latest konflux-test image: `quay.io/konflux-ci/konflux-test:1.5.2@sha256:c370a7bc6e172cdbab001f1a19f37ebc95f47c3d8fe2d915fa4641132490ead7`

File: pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml [hunk 2]
```diff
@@ -89,7 +89,85 @@ spec:
...
+              #!/usr/bin/env bash
+              set -euo pipefail
+
+              if [[ -z "${SNAPSHOT}" ]]; then
+                echo "SNAPSHOT parameter is required"
+                exit 1
+              fi
+
+              if [[ -z "${PACKAGE_NAME}" ]]; then
+                echo "WARNING: PACKAGE_NAME parameter is not set, using empty package name"
```
Comment (nmars) [inline pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml]: I am not sure if we should print `WARNING` here since it is functioning as designed. If no value is provided the step will discover the package name (or names) in the fragment.

Maybe phrase it like: "PACKAGE_NAME parameter is empty: the step will discover the package name in the FBC fragment"

File: pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml [hunk 3]
```diff
@@ -89,7 +89,85 @@ spec:
...
+              fi
+
+              if [[ -z "${CHANNEL_NAME}" ]]; then
+                echo "WARNING: CHANNEL_NAME parameter is not set, using empty channel name"
```
Comment (nmars) [inline pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml]: Same here. I suggest we phrase it like "CHANNEL_NAME parameter is empty. The step will determine the default channel name of the selected package"

File: pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml [hunk 4]
```diff
@@ -91,7 +91,85 @@ spec:
   workspaces:
     - name: shared-data
   tasks:
+    - name: validate-parameters
+      taskSpec:
+        steps:
+          - name: validate-parameters
+            image: quay.io/konflux-ci/konflux-test:v1.4.48@sha256:9b815268fb2bf10b5d745518da1c6568944f15816efe51adc192972b42a6e74d
```
Comment (nmars) [inline pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml]: Same as above. I suggest we use the latest konflux-test image.

File: pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml [hunk 5]
```diff
@@ -91,7 +91,85 @@ spec:
...
+            script: |
+              #!/usr/bin/env bash
+              set -euo pipefail
+
+              if [[ -z "${SNAPSHOT}" ]]; then
+                echo "SNAPSHOT parameter is required"
+                exit 1
+              fi
+
+              if [[ -z "${PACKAGE_NAME}" ]]; then
```
Comment (nmars) [inline pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml]: Same as above. I think we should not use "WARNING" and instead phrase it like "PACKAGE_NAME parameter is empty: the step will discover the package name in the FBC fragment"

File: pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml [hunk 6]
```diff
@@ -91,7 +91,85 @@ spec:
...
+      runAfter:
+        - validate-parameters
+      when:
+        - input: $(tasks.parse-metadata.results.test-event-type)
+          operator: notin
+          values: ["push", "Push"]
+      taskSpec:
+        steps:
+          - name: log-skipped-tasks
+            image: quay.io/konflux-ci/konflux-test:v1.4.48@sha256:9b815268fb2bf10b5d745518da1c6568944f15816efe51adc192972b42a6e74d
```
Comment (nmars) [inline pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml]: Same as above, I think we should move all instances of konflux-test to the latest v1.5.52.

**Review by nmars — COMMENTED** [review]
Also tagging @asergienk for her opinion.

File: pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml [hunk 7]
```diff
@@ -91,7 +91,85 @@ spec:
...
+              fi
+
+              if [[ -z "${SLACK_KEY_NAME}" ]]; then
+                echo "WARNING: SLACK_KEY_NAME parameter is not set, sending notifications to Slack may not work."
+              fi
+    - name: log-non-push-tasks
+      runAfter:
+        - validate-parameters
+      when:
+        - input: $(tasks.parse-metadata.results.test-event-type)
```
Comment (asergienk) [inline pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml:150]: This line is trying to read the result of `parse-metadata` task that did not run yet, `parse-metadata` task has to come before `log-non-push-tasks` to evaluate it's result

**Comment by asergienk** [comment]
Could you please update the README file to add new tasks: https://github.com/JAVGan/tekton-integration-catalog/blob/5a7cc1e805d7b4af3a1542e1d7e214e6062d90f9/pipelines/deploy-fbc-operator/0.2/README.MD

File: pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml (hunk 4 above)
Comment (JAVGan) [reply pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml]: Patched on c476922ea98ddd65ccc830e08b3c159218ac3bc0

File: pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml (hunk 6 above)
Comment (JAVGan) [reply pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml]: Patched on c476922ea98ddd65ccc830e08b3c159218ac3bc0

File: pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml (hunk 2 above)
Comment (JAVGan) [reply pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml]: Patched on 79f9fb17428a93343487585f4e762cdd5deb7a0f

File: pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml (hunk 5 above)
Comment (JAVGan) [reply pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml]: Patched on 79f9fb17428a93343487585f4e762cdd5deb7a0f

File: pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml (hunk 3 above)
Comment (JAVGan) [reply pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml]: Patched on 79f9fb17428a93343487585f4e762cdd5deb7a0f

File: pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml (hunk 1 above)
Comment (JAVGan) [reply pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml]: Patched on c476922ea98ddd65ccc830e08b3c159218ac3bc0

File: pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml (hunk 7 above)
Comment (JAVGan) [reply pipelines/deploy-fbc-operator/0.2/deploy-fbc-operator.yaml:150]: Thanks! I've patched it on e81d952525f0fca35b8a098f6639432b606fdbc1

**Comment by JAVGan** [comment]
@nmars @asergienk can you please check again?

File: pipelines/deploy-fbc-operator/0.1/README.MD [hunk 8]
```diff
@@ -4,25 +4,32 @@ The deploy-fbc-operator pipeline automates the provisioning of a EaaS Hypershift
 ## Pipeline Flow
 The pipeline consists of the following tasks:
 
-1. **Parse Metadata (`parse-metadata`)**  
+1. **Validate Parameters (`validate-parameters`)**
```
Comment (asergienk) [inline pipelines/deploy-fbc-operator/0.1/README.MD:7]: I think there has to be an extra new line after this, right now the tw lines appear together for #1 and #3:
https://github.com/JAVGan/tekton-integration-catalog/blob/938ce200e6c4786af606df27b93484827571631a/pipelines/deploy-fbc-operator/0.1/README.MD

File: pipelines/deploy-fbc-operator/0.1/README.MD [hunk 9]
```diff
@@ -4,25 +4,32 @@ The deploy-fbc-operator pipeline automates the provisioning of a EaaS Hypershift
...
+   Validates whether all required parameters are properly passed to the pipeline.
+
+2. **Parse Metadata (`parse-metadata`)**  
    Extracts metadata from the provided *snapshot*, including:
      * FBC fragment container image
      * Source Git URL
      * Git revision
 
-2. **Provision EAAS Space (`provision-eaas-space`)**  
+3. **Log Non Push Tasks `log-non-push-tasks`**
```
Comment (asergienk) [inline pipelines/deploy-fbc-operator/0.1/README.MD]: nit: missing parenthesis around `log-non-push-tasks` to match the style of the doc

**Comment by asergienk** [comment]
@JAVGan just small comments, otherwise lgtm! Was this PR tested on a cluster?

File: pipelines/deploy-fbc-operator/0.1/README.MD (hunk 8 above)
Comment (JAVGan) [reply pipelines/deploy-fbc-operator/0.1/README.MD:7]: Ok, I can fix it

File: pipelines/deploy-fbc-operator/0.1/README.MD (hunk 9 above)
Comment (JAVGan) [reply pipelines/deploy-fbc-operator/0.1/README.MD]: Ack, I'll fix it

**Comment by JAVGan** [comment]
> @JAVGan just small comments, otherwise lgtm! Was this PR tested on a cluster?

No, it was not yet.

@nmars can you please help me on testing this? I'll push some small fixes on README but the core logic will be the same

File: pipelines/deploy-fbc-operator/0.1/README.MD (hunk 9 above)
Comment (JAVGan) [reply pipelines/deploy-fbc-operator/0.1/README.MD]: Fixed

File: pipelines/deploy-fbc-operator/0.1/README.MD (hunk 8 above)
Comment (JAVGan) [reply pipelines/deploy-fbc-operator/0.1/README.MD:7]: Fixed

**Review by asergienk — APPROVED** [review]

**Review by asergienk — CHANGES_REQUESTED** [review]
@ashwgit could you please address the comments from Qodo? If some of them are no applicable and you think they shouldn't be implemented could you please make a comment about that here? Other than that, everything looks good.

File: pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml [hunk 10]
```diff
@@ -89,7 +89,56 @@ spec:
...
+
+              if [[ -z "${params.CREDENTIALS_SECRET_NAME}" ]]; then
+                echo "CREDENTIALS_SECRET_NAME parameter is required"
+                exit 1
+              fi
+
+              if [[ -z "${params.OCI_REF}" ]]; then
+                echo "OCI_REF parameter is required"
+                exit 1
+              fi
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml]: @asergienk do we want to turn CREDENTIALS_SECRET_NAME and OCI_REF optional or we keep these as required ?

File: pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml [hunk 11]
```diff
@@ -89,7 +89,46 @@ spec:
...
   tasks:
+    - name: validate-parameters
+      taskSpec:
+        steps:
+          - name: validate-parameters
+            image: quay.io/konflux-ci/konflux-test:1.5.2@sha256:c370a7bc6e172cdbab001f1a19f37ebc95f47c3d8fe2d915fa4641132490ead7
+            script: |
+              #!/usr/bin/env bash
+              set -euo pipefail
+
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml:100]: I don't see any issue with short form.

**Comment by ashwgit** [comment]
@asergienk here's the passing test cases - 
1. version 0.1 - https://konflux-ui.apps.stone-stg-rh01.l2vh.p1.openshiftapps.com/ns/iib-tenant/applications/ashwkuma-app/pipelineruns/tekton-integration-catalog-4g9fr?integrationTestName=tekton-integration-catalog
2. version 0.2 - https://konflux-ui.apps.stone-stg-rh01.l2vh.p1.openshiftapps.com/ns/iib-tenant/applications/ashwkuma-app/pipelineruns/deploy-fbc-operator-run-b945x?integrationTestName=integration-test-2

Do let me know if you need anything else before we can merge it.

File: pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (ashwgit) [reply pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml]: if this is required, I would be creating a separate MR for adding this change.

[the remark this answers is not in the document]
Comment (ashwgit) [reply pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml]: @asergienk ^

**Review by asergienk — APPROVED** [review]

File: pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (asergienk) [reply pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml]: this parameters are required, without them we won't be able to pushstore artifacts

[the remark this answers is not in the document]
Comment (asergienk) [reply pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml]: I don't think a follow up PR is needed. This pipeline is not anticipated to run on any other even type other than `push`.

**Review by nmars — APPROVED** [review]

**Comment by ashwgit** [comment]
> ## Review
> ### Findings
> #### Low
> 
>     * **[injection]** `pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml` — The `validate-parameters` task interpolates pipeline parameters directly into a bash `script:` block via `$(params.*)`. While the attack surface is minimal (parameters come from PAC automation, not end-user input), defense-in-depth recommends using `env:` blocks, consistent with the pattern used by other steps in the same pipeline (e.g., `retrieve-auth-header`). Same pattern in 0.2.
> 
>     * **[injection]** `pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml` — The `log-non-push-tasks` task interpolates `$(tasks.parse-metadata.results.test-event-type)` directly into a bash `echo` statement. Same recommendation: use an `env:` block for defense-in-depth. Same pattern in 0.2.
> 
>     * **[edge-case]** `pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml` — The `validate-parameters` task does not validate the `SNAPSHOT` parameter, which is required (no default) and is the primary input to `parse-metadata`. Tekton rejects missing required params, but an explicitly empty string would reach `parse-metadata` without a clear validation error. Same applies to 0.2.
> 
>     * **[error-message-format]** `pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml` — Inconsistent message prefixes in `validate-parameters`: warning messages use `WARNING:` prefix, but fatal errors before `exit 1` have no prefix. Consider adding `ERROR:` prefix and writing to stderr for consistency. Same in 0.2.
> 
>     * **[maintainability]** `pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml` — The `log-non-push-tasks` script contains a hardcoded list of 7 task names matching the pipeline's push-gated tasks. This list must be manually kept in sync if tasks are added or removed. Same in 0.2.
> 
>     * **[intent-category-mismatch]** PR title uses `fix(CLOUDDST-30777)` but the changes add new functionality (validation and logging tasks). `feat` or `chore` would more accurately categorize this change.
> 
>     * **[script-idiom]** `pipelines/deploy-fbc-operator/0.1/deploy-fbc-operator.yaml` — The `log-non-push-tasks` script uses a bash array (`skipped_tasks=(...)`) not found elsewhere in codebase inline scripts. Functional but introduces a new idiom. Same in 0.2.
> 
> 
> **Labels:** PR adds new pipeline tasks and bumps image versions in the deploy-fbc-operator pipeline

-------
Injection Issues - 
-  As the bash commands does not directly comes from user, we can skip this improvement or we can create a separate MR for improving injection related issues. 

Edge Case for Snapshot Validation
-  I expect snapshots are created after application build and integration test are executed after application build , we expect to always have a snapshot. 

Error-message-format 
-  For the error part, fail fast is more suitable , so I don't see any use of "ERROR" prefix. 

Intent-category-mismatch
- The title is more obvious to the fix we are developing.

**Comment by ashwgit** [comment]
@maneeshmehra I have provided comments for the reviews by fullsend, please have a look and let me know if we need any further improvement for this MR.

**Comment by maneeshmehra** [comment]
Approved.

**Comment by ashwgit** [comment]
@konflux-ci/konflux-devprod-team could you please have a look on this MR please?

**Review by Dannyb48 — APPROVED** [review]
LGTM
